cluster/engine.go

```diff
@@ -104,18 +106,53 @@ func (e *Engine) ConnectWithClient(client dockerclient.Client) error {
 	go e.refreshLoop()
 
 	// Start monitoring events from the engine.
-	e.client.StartMonitorEvents(e.handler, nil)
+	go e.startMonitorEvents()
 	e.emitEvent("engine_connect")
 
 	return nil
 }
 
+func (e *Engine) startMonitorEvents() {
+	stopCh := make(chan struct{})
+	ch, err := e.client.MonitorEvents(nil, stopCh)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"engine_id": e.ID,
+			"engine_ip": e.IP,
+		}).Warning("Start event monitor failed")
+		e.healthy = false
+		return
+	}
+	for {
+		select {
+		case ev := <-ch:
+			if ev.Error == nil {
+				go e.handler(&ev.Event)
+			} else {
+				log.WithFields(log.Fields{
+					"engine_id": e.ID,
+					"engine_ip": e.IP,
+				}).Warning("Receive event failed: ", ev.Error)
+				// Perhaps we can set e.healthy = false here?
+				return
+			}
+		case <-e.stopMonitorEventsCh:
+			stopCh <- struct{}{}
+			return
+		}
+	}
+}
+
+func (e *Engine) stopMonitorEvents() {
+	e.stopMonitorEventsCh <- struct{}{}
+}
+
 // Disconnect will stop all monitoring of the engine.
 // The Engine object cannot be further used without reconnecting it first.
 func (e *Engine) Disconnect() {
 	// do not close the chan, so it wait until the refreshLoop goroutine stops
 	e.stopCh <- struct{}{}
-	e.client.StopAllMonitorEvents()
+	e.stopMonitorEvents()
```

Comment:
I think we should stop monitoring events before stopping the refresh loop

WDYT?